### PR TITLE
ENH: sparse: avoid setitem densification

### PR DESCRIPTION
**Pull request by perimosocordiae**
Previously, assigning a sparse matrix would densify it. This had some unfortunate side-effects:

``` python
a = csr_matrix(np.eye(3))
# a.nnz == 3
a[0] = a[0]
# a.nnz == 5
```

This PR avoids densifying sparse arguments and improves the tests to make sure the previous behavior resulted in failure.

### CODE REVIEW

File: scipy/sparse/compressed.py [hunk 1]
```diff
@@ -614,14 +614,32 @@ def __setitem__(self, index, x):
         i, j = self._index_to_arrays(i, j)
 
         if isspmatrix(x):
-            x = x.toarray()
-
-        # Make x and i into the same shape
-        x = np.asarray(x, dtype=self.dtype)
-        x, _ = np.broadcast_arrays(x, i)
+            broadcast_row = x.shape[0] == 1 and i.shape[0] != 1
```
Comment (pv) [inline scipy/sparse/compressed.py]: What if `i.ndim != 2`?

Comment (pv) [inline scipy/sparse/compressed.py]: Hmm, probably can't happen due to `_index_to_arrays` (I don't remember any more exactly how all this worked :)

Comment (perimosocordiae) [inline scipy/sparse/compressed.py]: Yep, `_index_to_arrays` [makes sure that `i` is two-dimensional](https://github.com/perimosocordiae/scipy/blob/dfcf92e9d094f331c136618009c4c3fd78aff9ab/scipy/sparse/sputils.py#L381). I'm not a huge fan of the current system of converting all indexing into full integer index arrays, but it does make the logic here a lot simpler.

**Comment by perimosocordiae** [comment]
It would be nice to get this into 0.18, if possible. This behavior was causing some very subtle bugs / performance issues.

**Comment by pv** [comment]
I think this is not correct since it does not assign implicit zero elements.
```
>>> import numpy as np
>>> from scipy.sparse import csr_matrix
>>> A = csr_matrix(np.random.rand(6,6))
>>> B = csr_matrix(np.eye(6))
>>> A[:,:] = B
>>> abs(A - B).max()
0.99174595724650816
```

**Comment by pv** [comment]
Also needs explicit tests --- probably not enough to rely on the existing ones.

**Comment by perimosocordiae** [comment]
@pv: Good point, that case was indeed failing and wasn't caught by the test suite. I just pushed a fix and a test to make sure we catch that problem in the future.

**Comment by rgommers** [comment]
The CI failure isn't relevant. Looks like this is ready to merge. @pv?

**Comment by rgommers** [comment]
> ``#6163`` Previously, assigning a sparse matrix would densify it.

@perimosocordiae can you find back the actual issue? The indicated one is a ``stats`` PR.

**Comment by ev-br** [comment]
I'm going to merge this now for 0.18.0. A reference, https://github.com/scipy/scipy/pull/6193#issuecomment-227019862, would still be nice.
Thanks @perimosocordiae, all.

**Comment by perimosocordiae** [comment]
@rgommers Oops, not sure why that PR was indicated in this one. As far as I know, there aren't any outstanding issues that this should close.

**Pull request merged**